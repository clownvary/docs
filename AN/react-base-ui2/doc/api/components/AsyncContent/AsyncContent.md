## AsyncContent 
---
AsyncContent is an UI component that can display content in waiting and show manner.

### Prop Types
Property | Type | Default Value | Is Required | Description
:--- | :--- | :--- | :--- | :---
data-qa-id|string|'AsyncContent'|false|Defines the unique id for automation test
prefix|string|`${DefaultCSSPrefix}-`|false|Determines the prefix for CSS class names
component|node|&ensp;|false|Async component will be rendered
loader|func|() => Promise.resolve()|false|Contains the job need be executed.
placeHolder|string|'loading'|false|The default charactor when loading.
placeHolderType|enum:<br>[PlaceHolderType.TEXT, PlaceHolderType.ICON]|PlaceHolderType.TEXT|false|The type of Placeholder.