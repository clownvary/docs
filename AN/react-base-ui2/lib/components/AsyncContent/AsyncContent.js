'use strict';Object.defineProperty(exports, "__esModule", { value: true });var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);var _createClass2 = require('babel-runtime/helpers/createClass');var _createClass3 = _interopRequireDefault(_createClass2);var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);var _inherits2 = require('babel-runtime/helpers/inherits');var _inherits3 = _interopRequireDefault(_inherits2);var _promise = require('babel-runtime/core-js/promise');var _promise2 = _interopRequireDefault(_promise);var _react = require('react');var _react2 = _interopRequireDefault(_react);
var _propTypes = require('prop-types');var _propTypes2 = _interopRequireDefault(_propTypes);
var _consts = require('../../consts');
var _createAsyncContent = require('./createAsyncContent');var _createAsyncContent2 = _interopRequireDefault(_createAsyncContent);
var _PlaceHolderType = require('./consts/PlaceHolderType');var PlaceHolderType = _interopRequireWildcard(_PlaceHolderType);function _interopRequireWildcard(obj) {if (obj && obj.__esModule) {return obj;} else {var newObj = {};if (obj != null) {for (var key in obj) {if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key];}}newObj.default = obj;return newObj;}}function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}

/**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          * Default Props for AsyncContent
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          */
var AsyncContentProps = {
  /**
                           * Defines the unique id for automation test
                           */
  'data-qa-id': 'AsyncContent',
  /**
                                 * Determines the prefix for CSS class names
                                 */
  prefix: _consts.DefaultCSSPrefix + '-',
  loader: function loader() {return _promise2.default.resolve();},
  /**
                                                                    * The default charactor when loading.
                                                                    */
  placeHolder: 'loading',
  /**
                           * The type of Placeholder.
                           */
  placeHolderType: PlaceHolderType.TEXT };

var AsyncContentPropTypes = {
  /**
                               * Defines the unique id for automation test
                               */
  'data-qa-id': _propTypes2.default.string,
  /**
                                             * Determines the prefix for CSS class names
                                             */
  prefix: _propTypes2.default.string,
  /**
                                       * Async component will be rendered
                                       */
  component: _propTypes2.default.node,
  /**
                                        * Contains the job need be executed.
                                        */
  loader: _propTypes2.default.func,
  /**
                                     * The default charactor when loading.
                                     */
  placeHolder: _propTypes2.default.string,
  /**
                                            * The type of Placeholder.
                                            */
  placeHolderType: _propTypes2.default.oneOf([
  PlaceHolderType.TEXT, PlaceHolderType.ICON]) };


/** AsyncContent is an UI component that can display content in waiting and show manner.*/var
AsyncContent = function (_React$PureComponent) {(0, _inherits3.default)(AsyncContent, _React$PureComponent);function AsyncContent() {(0, _classCallCheck3.default)(this, AsyncContent);return (0, _possibleConstructorReturn3.default)(this, (AsyncContent.__proto__ || (0, _getPrototypeOf2.default)(AsyncContent)).apply(this, arguments));}(0, _createClass3.default)(AsyncContent, [{ key: 'render',


    /** The default Props*/value: function render()


    {var _props =






      this.props,placeHolder = _props.placeHolder,loader = _props.loader,prefix = _props.prefix,component = _props.component,dataQAId = _props['data-qa-id'],placeHolderType = _props.placeHolderType;
      var newComponent = (0, _createAsyncContent2.default)(function () {return loader().then(function () {return component;});},
      { placeHolder: placeHolder, prefix: prefix, 'data-qa-id': dataQAId, placeHolderType: placeHolderType });
      return _react2.default.createElement(newComponent);
    } /** The display name*/ }]);return AsyncContent;}(_react2.default.PureComponent);AsyncContent.displayName = 'AsyncContent';AsyncContent.defaultProps = AsyncContentProps;AsyncContent.propTypes = AsyncContentPropTypes;exports.default =

AsyncContent;module.exports = exports['default'];