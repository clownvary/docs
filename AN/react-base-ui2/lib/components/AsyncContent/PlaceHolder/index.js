'use strict';Object.defineProperty(exports, "__esModule", { value: true });var _text = require('./text');var _text2 = _interopRequireDefault(_text);
var _icon = require('./icon');var _icon2 = _interopRequireDefault(_icon);function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}exports.default =

{
  textPlaceHolder: _text2.default,
  iconPlaceHolder: _icon2.default };module.exports = exports['default'];