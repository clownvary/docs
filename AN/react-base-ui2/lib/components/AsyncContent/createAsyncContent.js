'use strict';Object.defineProperty(exports, "__esModule", { value: true });var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);var _createClass2 = require('babel-runtime/helpers/createClass');var _createClass3 = _interopRequireDefault(_createClass2);var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);var _inherits2 = require('babel-runtime/helpers/inherits');var _inherits3 = _interopRequireDefault(_inherits2);var _react = require('react');var _react2 = _interopRequireDefault(_react);
var _isFunction = require('lodash/isFunction');var _isFunction2 = _interopRequireDefault(_isFunction);

var _consts = require('../../consts');
var _utils = require('../../utils');
var _PlaceHolder = require('./PlaceHolder');
var _PlaceHolderType = require('./consts/PlaceHolderType');var PlaceHolderType = _interopRequireWildcard(_PlaceHolderType);function _interopRequireWildcard(obj) {if (obj && obj.__esModule) {return obj;} else {var newObj = {};if (obj != null) {for (var key in obj) {if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key];}}newObj.default = obj;return newObj;}}function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}var

DATA_QA_ID = _consts.CustomAttr.DATA_QA_ID;exports.default =

function (loader, _ref) {var placeHolder = _ref.placeHolder,prefix = _ref.prefix,dataQAID = _ref[DATA_QA_ID],placeHolderType = _ref.placeHolderType;return function (_React$Component) {(0, _inherits3.default)(asyncContent, _React$Component);

    function asyncContent() {(0, _classCallCheck3.default)(this, asyncContent);var _this = (0, _possibleConstructorReturn3.default)(this, (asyncContent.__proto__ || (0, _getPrototypeOf2.default)(asyncContent)).call(this));

      _this.state = {
        component: null };


      _this.mounting = true;return _this;
    }(0, _createClass3.default)(asyncContent, [{ key: 'componentDidMount', value: function componentDidMount()

      {var _this2 = this;
        if (!(0, _isFunction2.default)(loader) && !(0, _utils.isPromise)(loader)) {
          return console.error('asyncContent: Loader is not function or promise');
        }

        var component = (0, _isFunction2.default)(loader) ? loader() : loader;

        if (!(0, _utils.isPromise)(component)) {
          return console.error('asyncContent: Loader doesn\'t return a promise');
        }

        return component.
        then(function (module) {return _this2.mounting && _this2.setState({ component: module });}).
        catch(function (err) {return console.error('asyncContent: Loading failed', err);});
      } }, { key: 'componentWillUnmount', value: function componentWillUnmount()

      {
        this.mounting = false;
      } }, { key: 'renderLoader', value: function renderLoader()

      {
        switch (placeHolderType) {
          case PlaceHolderType.ICON:
            return (0, _PlaceHolder.iconPlaceHolder)(placeHolder);
          case PlaceHolderType.TEXT:
          default:
            return (0, _PlaceHolder.textPlaceHolder)(placeHolder);}

      } }, { key: 'render', value: function render()

      {var
        component = this.state.component;

        return (
          _react2.default.createElement('div', {
              className: prefix + 'asynccontent',
              'data-qa-id': dataQAID },

            component || this.renderLoader()));

      } }]);return asyncContent;}(_react2.default.Component);};module.exports = exports['default'];