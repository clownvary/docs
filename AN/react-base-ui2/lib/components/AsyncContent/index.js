'use strict';Object.defineProperty(exports, "__esModule", { value: true });exports.PlaceHolderType = exports.createAsyncContent = exports.AsyncContent = undefined;var _AsyncContent = require('./AsyncContent');var _AsyncContent2 = _interopRequireDefault(_AsyncContent);
var _createAsyncContent = require('./createAsyncContent');var _createAsyncContent2 = _interopRequireDefault(_createAsyncContent);
var _PlaceHolderType = require('./consts/PlaceHolderType');var PlaceHolderType = _interopRequireWildcard(_PlaceHolderType);function _interopRequireWildcard(obj) {if (obj && obj.__esModule) {return obj;} else {var newObj = {};if (obj != null) {for (var key in obj) {if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key];}}newObj.default = obj;return newObj;}}function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}exports.default =

_AsyncContent2.default;exports.

AsyncContent = _AsyncContent2.default;exports.
createAsyncContent = _createAsyncContent2.default;exports.
PlaceHolderType = PlaceHolderType;