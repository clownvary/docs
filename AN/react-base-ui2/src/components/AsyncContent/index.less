/* stylelint-disable */
.an-asynccontent {
  display: inline;

  .text-loader {
    display: inline;

    &__text {
      padding-right: 2px;
    }

    @keyframes blink {
      30% {
        color: transparent;
      }
    }

    &__dot {
      animation: 1s blink infinite;
    }

    &__dot:nth-child(2) {
      animation-delay: 0.5s;
    }

    &__dot:nth-child(3) {
      animation-delay: 1s;
    }
  }

  .icon-loader {
    position: absolute;
    display: inline-flex;
    width: 100%;
    height: 100%;

    &__text {
      padding-left: 3px;
    }

    .icon-loading-m {
      padding: 1px; // fix the spin icon cutt issue in IE11

      font-size: 16px;
      color: #ffb245;
      transform-origin: center 10px; // fix the spin icon cutt issue in IE11
    }

    @keyframes icon-spin {
      to {
        transform: rotate(360deg);
      }
    }

    .icon-spin {
      animation: icon-spin 2s infinite linear;
    }
  }
}
