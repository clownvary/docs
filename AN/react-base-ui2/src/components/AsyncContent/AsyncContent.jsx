import React from 'react';
import PropTypes from 'prop-types';
import { DefaultCSSPrefix } from '../../consts';
import createAsyncContent from './createAsyncContent';
import * as PlaceHolderType from './consts/PlaceHolderType';

/**
 * Default Props for AsyncContent
 */
const AsyncContentProps = {
  /**
   * Defines the unique id for automation test
   */
  'data-qa-id': 'AsyncContent',
  /**
   * Determines the prefix for CSS class names
   */
  prefix: `${DefaultCSSPrefix}-`,
  loader: () => Promise.resolve(),
  /**
   * The default charactor when loading.
   */
  placeHolder: 'loading',
  /**
   * The type of Placeholder.
   */
  placeHolderType: PlaceHolderType.TEXT
};
const AsyncContentPropTypes = {
  /**
   * Defines the unique id for automation test
   */
  'data-qa-id': PropTypes.string,
  /**
   * Determines the prefix for CSS class names
   */
  prefix: PropTypes.string,
  /**
   * Async component will be rendered
   */
  component: PropTypes.node,
  /**
   * Contains the job need be executed.
   */
  loader: PropTypes.func,
  /**
   * The default charactor when loading.
   */
  placeHolder: PropTypes.string,
  /**
   * The type of Placeholder.
   */
  placeHolderType: PropTypes.oneOf([
    PlaceHolderType.TEXT, PlaceHolderType.ICON
  ])
};
/** AsyncContent is an UI component that can display content in waiting and show manner.*/
class AsyncContent extends React.PureComponent {
  /** The display name*/
  static displayName = 'AsyncContent';
  /** The default Props*/
  static defaultProps = AsyncContentProps;
  static propTypes = AsyncContentPropTypes;
  render() {
    const {
      placeHolder,
      loader,
      prefix,
      component,
      'data-qa-id': dataQAId,
      placeHolderType } = this.props;
    const newComponent = createAsyncContent(() => loader().then(() => component),
      { placeHolder, prefix, 'data-qa-id': dataQAId, placeHolderType });
    return React.createElement(newComponent);
  }
}
export default AsyncContent;
