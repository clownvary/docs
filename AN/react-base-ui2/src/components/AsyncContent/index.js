import AsyncContent from './AsyncContent';
import createAsyncContent from './createAsyncContent';
import * as PlaceHolderType from './consts/PlaceHolderType';

export default AsyncContent;
export {
  AsyncContent,
  createAsyncContent,
  PlaceHolderType
};
