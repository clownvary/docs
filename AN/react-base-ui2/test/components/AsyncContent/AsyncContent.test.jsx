import React from 'react';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import AsyncContent, { PlaceHolderType, createAsyncContent } from 'src/components/AsyncContent';


const component = <div>test</div>;

describe('src/components/AsyncContent', () => {
  test('rendering', () => {
    const wrapper = mount(<AsyncContent component={component} />);
    expect(toJSON(wrapper)).toMatchSnapshot();

    wrapper.setProps({ placeHolderType: PlaceHolderType.ICON });
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  test('createAsyncContent', () => {
    let AsyncComponent = createAsyncContent(
      null,
      {
        placeHolder: 'loading',
        prefix: 'an',
        'data-qa-id': 'test',
        placeHolderType: PlaceHolderType.TEXT
      }
    );

    let wrapper = mount(<AsyncComponent />);
    expect(toJSON(wrapper)).toMatchSnapshot();


    AsyncComponent = AsyncComponent = createAsyncContent(
      () => null,
      {
        placeHolder: 'loading',
        prefix: 'an',
        'data-qa-id': 'test',
        placeHolderType: PlaceHolderType.TEXT
      }
    );
    wrapper = mount(<AsyncComponent />);
    expect(toJSON(wrapper)).toMatchSnapshot();

    AsyncComponent = AsyncComponent = createAsyncContent(
      Promise.reject(),
      {
        placeHolder: 'loading',
        prefix: 'an',
        'data-qa-id': 'test',
        placeHolderType: PlaceHolderType.TEXT
      }
    );
    wrapper = mount(<AsyncComponent />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
