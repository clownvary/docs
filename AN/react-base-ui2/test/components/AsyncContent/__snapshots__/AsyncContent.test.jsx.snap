// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/AsyncContent createAsyncContent 1`] = `
<asyncContent>
  <div
    className="anasynccontent"
    data-qa-id="test"
  >
    <div
      className="text-loader"
    >
      <span
        className="text-loader__text"
      >
        loading
      </span>
      <span
        className="text-loader__dot"
      >
        .
      </span>
      <span
        className="text-loader__dot"
      >
        .
      </span>
      <span
        className="text-loader__dot"
      >
        .
      </span>
    </div>
  </div>
</asyncContent>
`;

exports[`src/components/AsyncContent createAsyncContent 2`] = `
<asyncContent>
  <div
    className="anasynccontent"
    data-qa-id="test"
  >
    <div
      className="text-loader"
    >
      <span
        className="text-loader__text"
      >
        loading
      </span>
      <span
        className="text-loader__dot"
      >
        .
      </span>
      <span
        className="text-loader__dot"
      >
        .
      </span>
      <span
        className="text-loader__dot"
      >
        .
      </span>
    </div>
  </div>
</asyncContent>
`;

exports[`src/components/AsyncContent createAsyncContent 3`] = `
<asyncContent>
  <div
    className="anasynccontent"
    data-qa-id="test"
  >
    <div
      className="text-loader"
    >
      <span
        className="text-loader__text"
      >
        loading
      </span>
      <span
        className="text-loader__dot"
      >
        .
      </span>
      <span
        className="text-loader__dot"
      >
        .
      </span>
      <span
        className="text-loader__dot"
      >
        .
      </span>
    </div>
  </div>
</asyncContent>
`;

exports[`src/components/AsyncContent rendering 1`] = `
<AsyncContent
  component={
    <div>
      test
    </div>
  }
  data-qa-id="AsyncContent"
  loader={[Function]}
  placeHolder="loading"
  placeHolderType="text"
  prefix="an-"
>
  <asyncContent>
    <div
      className="an-asynccontent"
      data-qa-id="AsyncContent"
    >
      <div
        className="text-loader"
      >
        <span
          className="text-loader__text"
        >
          loading
        </span>
        <span
          className="text-loader__dot"
        >
          .
        </span>
        <span
          className="text-loader__dot"
        >
          .
        </span>
        <span
          className="text-loader__dot"
        >
          .
        </span>
      </div>
    </div>
  </asyncContent>
</AsyncContent>
`;

exports[`src/components/AsyncContent rendering 2`] = `
<AsyncContent
  component={
    <div>
      test
    </div>
  }
  data-qa-id="AsyncContent"
  loader={[Function]}
  placeHolder="loading"
  placeHolderType="icon"
  prefix="an-"
>
  <asyncContent>
    <div
      className="an-asynccontent"
      data-qa-id="AsyncContent"
    >
      <div
        className="icon-loader"
      >
        <div
          className="icon-loader__icon"
        >
          <i
            className="icon icon-loading-m icon-spin"
          />
          <div
            className="icon-loader__text"
          >
            loading
            ...
          </div>
        </div>
      </div>
    </div>
  </asyncContent>
</AsyncContent>
`;
